cmd/snap-repair/runner.go

```diff
@@ -67,24 +82,145 @@ func (r *Repair) SetStatus(status RepairStatus) {
 
 // Run executes the repair script leaving execution trail files on disk.
 func (r *Repair) Run() error {
-	// XXX initial skeleton...
 	// write the script to disk
-	rundir := filepath.Join(dirs.SnapRepairRunDir, r.BrandID(), r.RepairID())
+	rundir := r.RunDir()
 	err := os.MkdirAll(rundir, 0775)
 	if err != nil {
 		return err
 	}
+
 	script := filepath.Join(rundir, fmt.Sprintf("script.r%d", r.Revision()))
-	err = osutil.AtomicWriteFile(script, r.Body(), 0600, 0)
+	err = osutil.AtomicWriteFile(script, r.Body(), 0700, 0)
+	if err != nil {
+		return err
+	}
+
+	// the date may be broken so we use an additional counter
+	baseName := fmt.Sprintf("r%d", r.Revision())
+	logPath := filepath.Join(rundir, baseName+".output")
+	logf, err := os.OpenFile(logPath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+	defer logf.Close()
+
+	statusR, statusW, err := os.Pipe()
 	if err != nil {
 		return err
 	}
+	defer statusR.Close()
+	defer statusW.Close()
+
+	// run the script
+	env := os.Environ()
+	// we need to hardcode FD=3 because this is the FD after
+	// exec.Command() forked. there is no way in go currently
+	// to run something right after fork() in the child to
+	// know the fd. However because go will close all fds
+	// except the ones in "cmd.ExtraFiles" we are safe to set "3"
+	env = append(env, "SNAP_REPAIR_STATUS_FD=3")
+	env = append(env, "SNAP_REPAIR_RUN_DIR="+rundir)
+
+	workdir := filepath.Join(rundir, "work")
+	if err := os.MkdirAll(workdir, 0700); err != nil {
+		return err
+	}
+
+	cmd := exec.Command(script)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	cmd.Env = env
+	cmd.Dir = workdir
+	cmd.ExtraFiles = []*os.File{statusW}
+	cmd.Stdout = logf
+	cmd.Stderr = logf
+	if err = cmd.Start(); err != nil {
+		return err
+	}
+	statusW.Close()
 
-	// XXX actually run things and captures output etc
+	// wait for repair to finish or timeout
+	killTimerCh := time.After(defaultRepairTimeout)
+	doneCh := make(chan error)
+	go func() {
+		// ignore error, we only care about what we got via
+		// the status pipe
+		doneCh <- cmd.Wait()
+		close(doneCh)
+	}()
+	select {
+	case err = <-doneCh:
+		// done
+	case <-killTimerCh:
+		if err := osutil.KillProcessGroup(cmd); err != nil {
+			logger.Noticef("cannot kill timed out repair %s: %s", r, err)
+		}
+		err = fmt.Errorf("repair did not finish within %s", defaultRepairTimeout)
+	}
+	if err != nil {
+		err = fmt.Errorf("%q failed: %s", r.Ref(), err)
+
+		statusPath := filepath.Join(rundir, baseName+".retry")
```

Comment:
I think this changes what gustavo said about ignoring the error status from the POV of considering it done. To be clear I think we still want to report to the error tracker because it's weird.

This probably means first scanning the status stream and then consider the error to send thing to the tracker

Reply:
Thanks, I changed the code now so that its closer to what we had before. It will send errors to the error tracker but still honor the status that was send via the status-fd. Also script errors will not lead to run exiting with an error, instead it will just retry and report to the errtracker.